### PR TITLE
[dataset] Add a `reorder` flag to builtin datasets

### PR DESCRIPTION
**Pull request by hhr114**
## Description

Refer to issues https://github.com/dmlc/dgl/issues/3987 https://github.com/dmlc/dgl/issues/3271

Thx for your kind review. @yzh119 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes

- [x] Add argument `reorder=False` for citation_graph

### CODE REVIEW

**Comment by mufeili** [comment]
Could you also add the description of the argument `reorder`? Also it will be great if you could do that for the other datasets after we merge this PR.

**Comment by hhr114** [comment]
> Could you also add the description of the argument `reorder`? Also it will be great if you could do that for the other datasets after we merge this PR.

I have add the description of the argument `reorder`. How about the code now?

**Comment by hhr114** [comment]
> Could you also add the description of the argument `reorder`? Also it will be great if you could do that for the other datasets after we merge this PR.

I will set the `reorder` argument for other datasets in the near future. Btw, I find that `assert np.array_equal(dst, np.sort(dst))` in test_data.py cause an error. Should I delete that line?

**Review by yzh119 — CHANGES_REQUESTED** [review]
@hhr114 Thank you so much for the fix.

However, DGL would cache the dataset by default, we must change the cache behavior correspondingly.

More specifically, DGL saves the reordered and un-reordered graph under the same path, which may cause some confusion:
https://github.com/dmlc/dgl/blob/94b7199a2336ea68a5f79254792d1d66463aa90d/python/dgl/data/citation_graph.py#L180-L194

You can either save the reordered/un-reordered graph under different paths or move the reorder logic to `load` function instead of `process` function.

**Comment by hhr114** [comment]
@yzh119 You're right. I will fix it as you say.

File: python/dgl/data/citation_graph.py [hunk 1]
```diff
@@ -51,6 +51,8 @@ class CitationGraphDataset(DGLBuiltinDataset):
         A transform that takes in a :class:`~dgl.DGLGraph` object and returns
         a transformed version. The :class:`~dgl.DGLGraph` object will be
         transformed before every access.
+    reorder : bool
+        Whether to reorder the graph. Default: False.
```
Comment (mufeili) [inline python/dgl/data/citation_graph.py:55]: "Whether to reorder the graph" -> "Whether to reorder the graph using :func:`~dgl.reorder_graph`"

**Comment by mufeili** [comment]
> > Could you also add the description of the argument `reorder`? Also it will be great if you could do that for the other datasets after we merge this PR.
> 
> I have add the description of the argument `reorder`. How about the code now?

I left a minor comment.

**Comment by mufeili** [comment]
> > Could you also add the description of the argument `reorder`? Also it will be great if you could do that for the other datasets after we merge this PR.
> 
> I will set the `reorder` argument for other datasets in the near future. Btw, I find that `assert np.array_equal(dst, np.sort(dst))` in test_data.py cause an error. Should I delete that line?

Either you can remove that line or alternatively you can add a few additional lines of code for testing an ordered version with `force_reload=True`.

**Comment by mufeili** [comment]
> @hhr114 Thank you so much for the fix.
> 
> However, DGL would cache the dataset by default, we must change the cache behavior correspondingly.
> 
> More specifically, DGL saves the reordered and un-reordered graph under the same path, which may cause some confusion:
> 
> https://github.com/dmlc/dgl/blob/94b7199a2336ea68a5f79254792d1d66463aa90d/python/dgl/data/citation_graph.py#L180-L194
> 
> You can either save the reordered/un-reordered graph under different paths or move the reorder logic to `load` function instead of `process` function.

@yzh119 @hhr114 I suggest adding a version number postfix to the cached file name. In the future, we might have new breaking changes. It's safer to force reloading every time there is a new release of DGL.

**Comment by hhr114** [comment]
> @yzh119 @hhr114 I suggest adding a version number postfix to the cached file name. In the future, we might have new breaking changes. It's safer to force reloading every time there is a new release of DGL.

To add a version number postfix, I'm not sure if I can `import dgl` and call `dgl.__version__` in `dgl_dataset.py`.

**Comment by mufeili** [comment]
> > @yzh119 @hhr114 I suggest adding a version number postfix to the cached file name. In the future, we might have new breaking changes. It's safer to force reloading every time there is a new release of DGL.
> 
> To add a version number postfix, I'm not sure if I can `import dgl` and call `dgl.__version__` in `dgl_dataset.py`.

You should be able to do something like this `https://github.com/dmlc/dgl/blob/62e23bd5b14936dd471feeb1364c6b70eb9504d8/python/dgl/__init__.py#L32`

**Review by yzh119 — APPROVED** [review]
LGTM, I'll approve this PR.

@mufeili wikics and reddit have similiar issues and we can fix them in later PRs.

**Comment by mufeili** [comment]
> LGTM, I'll approve this PR.
> 
> @mufeili wikics and reddit have similiar issues and we can fix them in later PRs.

Sounds good.

**Pull request merged**